src/useImageDimensions.ts

```diff
@@ -0,0 +1,61 @@
+import {useEffect, useState} from 'react'
+import {Image, ImageRequireSource} from 'react-native'
+
+export interface URISource {
+  uri: string
+}
+
+/**
+ * @param source either a remote URL or a local file resource.
+ * @returns original image dimensions (width, height and aspect ratio).
+ */
+function useImageDimensions(source: ImageRequireSource | URISource) {
+  const [[dimensions, error], setState] = useState<
+    [{width: number; height: number}?, Error?]
+  >([])
+
+  useEffect(() => {
+    try {
+      if (typeof source === 'number') {
+        const {width, height} = Image.resolveAssetSource(source)
+        setState([{width, height}])
+      } else if (typeof source === 'object' && source.uri) {
+        setState([])
+        Image.getSize(
+          source.uri,
+          (width, height) => setState([{width, height}]),
+          e => setState([dimensions, e]),
+        )
+      } else {
+        throw new Error('not implemented')
+      }
+    } catch (e) {
+      setState([dimensions, e])
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [source])
+
+  return {
+    dimensions:
+      dimensions &&
+      (Object.setPrototypeOf(dimensions, {
+        get aspectRatio(): number {
+          const _this = this as any
+          return _this.width / _this.height
+        },
+      }) as {
+        /**
+         * width to height ratio
+         */
+        readonly aspectRatio: number
+        width: number
+        height: number
```

Comment:
How about making `width` and `height` `readonly` as well?

Reply:
Unlike 'aspectRatio', actually, these fields are not read-only.

Reply:
I think it could be good to mark them as readonly so catch a potential error where the end user accidentally modifies them, since that will be reset each render loop. What do you think?

Reply:
I don't understand what do you mean and why do you think that field change will cause anything?